### PR TITLE
refactor(repl): use SWC lexer to highlight and validate

### LINKED ISSUE #8160: Division operator gets mistaken for regex literals in multi-line REPL editor
The following hangs waiting for another line of input

```
Deno 1.5.0
exit using ctrl+d or close()
> (3 / 4)
```

Cause is quite simple, the validator believes that the `/` token is the start of a regex literal.

### LINKED ISSUE #8240: [CLI] REPL syntax highlighting doesn't work inside of template strings
Template string values are interpreted as usual strings in the REPL CLI:

Similar to #8202.

### PR DESCRIPTION
**Pull request by Liamolucko**
This more or less just swaps out manual parsing for the SWC lexer.

Fixes #8160. Fixes #8240.

### CODE REVIEW

**Comment by Liamolucko** [comment]
That was weird, `tools/lint.js` wasn't showing any issues for some reason.

File: cli/tools/repl.rs [hunk 1]
```diff
@@ -211,71 +214,98 @@ impl Highlighter for Helper {
...
+            Token::Word(word) => match word {
+              Word::True | Word::False | Word::Null => {
+                colors::yellow(&line[item.span]).to_string()
+              }
+              Word::Keyword(_) => colors::cyan(&line[item.span]).to_string(),
+              Word::Ident(js_word!("undefined")) => {
+                colors::gray(&line[item.span]).to_string()
+              }
+              Word::Ident(js_word!("Infinity"))
+              | Word::Ident(js_word!("NaN")) => {
```
Comment (bartlomieju) [inline cli/tools/repl.rs:295]: I think we could get away without using `js_word!` an instead calling `into()` on the strings

File: cli/Cargo.toml [hunk 2]
```diff
@@ -65,6 +65,7 @@ serde = { version = "1.0.116", features = ["derive"] }
 shell-escape = "0.1.5"
 sys-info = "0.7.0"
 sourcemap = "6.0.1"
+swc_atoms = "0.2.5"
```
Comment (bartlomieju) [inline cli/Cargo.toml:68]: Remove

File: cli/tools/repl.rs [hunk 3]
```diff
@@ -211,71 +214,98 @@ impl Highlighter for Helper {
...
+    struct Item {
+      span: Range<usize>,
+      kind: ItemKind,
+    }
+
+    fn span_to_range(span: Span) -> Range<usize> {
+      span.lo.0 as usize..span.hi.0 as usize
+    }
+
+    let mut items: Vec<Item> = Vec::new();
```
Comment (bartlomieju) [inline cli/tools/repl.rs:264]: I think this utility should be put in `cli/ast.rs` and could potenatially be reused for doc generator text output

**Review by bartlomieju — COMMENTED** [review]
@caspervonb please take a look

File: cli/tools/repl.rs (hunk 1 above)
Comment (Liamolucko) [reply cli/tools/repl.rs]: Ok, fixed - `into()` didn't work, but `Atom` implements `PartialEq` for `str`.

File: cli/tools/repl.rs [hunk 4]
```diff
@@ -133,55 +139,51 @@ impl Validator for Helper {
...
+        BytePos(0),
+        BytePos(ctx.input().len() as u32),
+      ),
+      None,
+    );
+
+    let mut stack: Vec<Token> = Vec::new();
+    let mut in_template = false;
+
+    for item in lexer {
```
Comment (caspervonb) [inline cli/tools/repl.rs:156]: LGTM

File: cli/tools/repl.rs [hunk 5]
```diff
@@ -211,71 +213,99 @@ impl Highlighter for Helper {
...
-        } else if let Some(cap) = caps.name("binarynumber") {
-          colors::yellow(cap.as_str()).to_string()
-        } else {
-          caps[0].to_string()
-        }
-      })
-      .to_string()
-      .into()
+    let comments = SingleThreadedComments::default();
+    let lexer = Lexer::new(
```
Comment (caspervonb) [inline cli/tools/repl.rs:236]: Clean, LGTM.

**Review by caspervonb — APPROVED** [review]
Look good to me, clean implementation 👍

File: cli/tools/repl.rs (hunk 3 above)
Comment (Liamolucko) [reply cli/tools/repl.rs]: I've created `ast::lex()` now, assuming that's what you meant.

**Review by caspervonb — CHANGES_REQUESTED** [review]
Oops, approved a bit too soon, seems that big integer literals aren't accounted for here.

**Comment by Liamolucko** [comment]
Whoops, that's fixed now.

File: Cargo.lock [hunk 6]
```diff
@@ -2350,9 +2350,9 @@ checksum = "a2eb9349b6444b326872e140eb1cf5e7c522154d69e7a0ffb0fb81c06b37543f"
 
 [[package]]
 name = "string_cache"
-version = "0.8.0"
+version = "0.8.1"
 source = "registry+https://github.com/rust-lang/crates.io-index"
-checksum = "2940c75beb4e3bf3a494cef919a747a2cb81e52571e212bfbd185074add7208a"
+checksum = "8ddb1139b5353f96e429e1a5e19fbaf663bddedaa06d1dbd49f82e352601209a"
```
Comment (bartlomieju) [inline Cargo.lock:2355]: What caused this update?

File: cli/tools/repl.rs (hunk 3 above)
Comment (bartlomieju) [reply cli/tools/repl.rs]: Yes, that's great, thanks

File: Cargo.lock (hunk 6 above)
Comment (bartlomieju) [reply Cargo.lock]: I've reverted this change.

**Review by bartlomieju — APPROVED** [review]
LGTM, nice cleanup @Liamolucko

**Comment by caspervonb** [comment]
Great contribution, thanks @Liamolucko 👏 🥳

**Pull request merged**